core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,455 @@
+                        return;
+                    }
+                    final Subscription upstream = this.upstream;
+                    if (upstream != null) {
+                       requestUpstream(upstream);
+                    }
+                    break;
+                }
+                case REQUIRE_TAIL: {
+                    sendTail();
+                    break;
+                }
+            }
+        }
+
+        private void sendHead() {
+            setState(State.REQUIRE_HEAD, State.REQUIRE_BODY);
+            assert head != null;
+            final T head = this.head;
+            this.head = null;
+            publishDownstream(head, true);
+        }
+
+        private void sendTail() {
+            assert state == State.REQUIRE_TAIL;
+            if (tail != null) {
+                final T tail = this.tail;
+                this.tail = null;
+                downstream.onNext(tail);
+            }
+            close0(null);
+        }
+
+        private void requestUpstream(Subscription subscription) {
+            if (requested <= 0) {
+                return;
+            }
+            assert upstreamRequested == 0;
+            upstreamRequested = requested;
+            if (requested < Long.MAX_VALUE) {
+                requested = 0;
+            }
+            subscription.request(upstreamRequested);
+        }
+
+        private void publishDownstream(T item, boolean head) {
+            requireNonNull(item, "item");
+            if (state == State.DONE) {
+                PooledObjects.close(item);
+                return;
+            }
+            downstream.onNext(item);
+
+            if (head) {
+                if (requested < Long.MAX_VALUE) {
+                    requested--;
+                }
+                subscribed = true;
+                publisher.subscribe(this, executor, options);
+            } else {
+                assert upstreamRequested > 0;
+                if (upstreamRequested < Long.MAX_VALUE) {
+                    upstreamRequested--;
+                }
+            }
+
+            if (!publishedAny) {
+                publishedAny = true;
+            }
+
+            publish();
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (state == State.DONE) {
+                subscription.cancel();
+                return;
+            }
+            upstream = subscription;
+            requestUpstream(subscription);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            publishDownstream(item, false);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            close(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            setState(State.REQUIRE_BODY, State.REQUIRE_TAIL);
```

Comment:
I think it's possible that the stream is closed before `onComplete` is called, so `state == DONE` when this line is called.

Can you add a check so that an error isn't thrown here?

```suggestion
            if (state == State.DONE) {
                return;
            }
            setState(State.REQUIRE_BODY, State.REQUIRE_TAIL);
```

Reply:
Reproducible by 

```java
final HttpRequest req =
        HttpRequest.of(RequestHeaders.of(HttpMethod.GET, "/"),
                       StreamMessage.of(HttpData.ofAscii("a")),
                       HttpHeaders.of("asdf", "asdf"));
StepVerifier.create(req, 3)
            .expectComplete()
            .verify();
```

Reply:
✅ updated!